content/en/blog/_posts/2021-04-15-local-storage-features-go-beta.md

```diff
@@ -0,0 +1,164 @@
+nodes. Because information might be not quite up-to-date, pods may
+still get assigned to nodes with insufficient storage, it's just less
+likely and the next scheduling attempt for a pod should work better
+once the information got refreshed.
+
+So CSI drivers still need the ability to recover from a bad scheduling
+decision. That was one the reasons for the introduction of "*generic*
+ephemeral volumes", the other feature that got promoted to beta in
+1.21. They are not limited to local storage and, as the name implies,
+also work for non-CSI storage drivers. The feature is based on
+automatically creating PVCs from on a template in the pod spec and
+then lettings storage drivers react to those as if the PVC had been
+created manually.
+
+## ... and the ugly.
+
+Both generic ephemeral volumes and storage capacity tracking increase
+the load on the API server. Whether that is a problem depends a lot on
+the kind of workload, in particular how many pods have volumes and how
+often those need to be created and destroyed.
+
+No attempt was made to model how scheduling decisions affect storage
+capacity. That's because the effect can vary considerably depending on
+how the storage system handles storage. The effect is that multiple
+pods with unbound volumes might get assigned to the same node even
+though capacity there is only sufficient for one pod. Scheduling
+should recover, but it would be more efficient if the scheduler knew
+more about storage.
+
+Because storage capacity gets published by a running CSI driver and
+the cluster autoscaler needs information about a node that hasn't been
+created yet, it will currently not scale up a cluster for pods that
+need volumes. There is an [idea how to provide that
+information](https://github.com/kubernetes/autoscaler/pull/3887), but
+more work is needed in that area.
+
+Distributed snapshotting and resizing are not currently supported. It
+should be doable to adapt the respective sidecar and there are
+tracking issues for external-snapshotter and external-resizer open
+already, they just need some volunteer.
+
+The recovery from a bad scheduling decising can fail for pods with
+multiple volumes, in particular when those volumes are local to nodes:
+if one volume can be created and then storage is insufficient for
+another volume, the first volume continues to exist and forces the
+scheduler to put the pod onto the node of that volume. There is an
+idea how do deal with this, [rolling back the provision of the
+volume](https://github.com/kubernetes/enhancements/pull/1703), but
+this is only in the very early stages of brainstorming and not even a
+merged KEP yet. For now it is better to avoid creating pods with more
+than one volume.
+
+## Enabling the new features and next steps
+
+Because the Kubernetes features are now beta, no changes are needed
+there anymore when deploying a Kubernetes 1.21 cluster. Generic
+ephemeral volumes also work without changes in CSI drivers.
```

Comment:
It is not clear to me how the Generic ephemeral volume feature can be used by itself and what is the benefit.
It's good to clarify what is the benefit of each of these two features when used separately or together.

Reply:
> It is not clear to me how the Generic ephemeral volume feature can be used by itself and what is the benefit.

It might be as simple as "previously you couldn't do ephemeral volumes with CSI driver XYZ, now you can"... Or were you wondering why someone might want to do that? One use case is local scratch space.

I mention that use case now when introducing ephemeral volumes and also linked to the previous blog post which had a more in-depth explanation.